pennylane/workflow/_setup_transform_program.py

```diff
@@ -0,0 +1,104 @@
+    that there is only one ``dynamic_one_shot`` transform in the outer and inner transform
+    programs combined.
+
+    """
+
+    all_transforms = outer_transform + inner_transform
+    type_to_keep = 0
+    if any("mid_circuit_measurements" in str(t) for t in all_transforms):
+        type_to_keep = 2
+    elif any("dynamic_one_shot" in str(t) for t in all_transforms):
+        type_to_keep = 1
+
+    if type_to_keep == 0:
+        return
+
+    dynamic_transform_found = inner_transform.prune_dynamic_transform(type_to_keep)
+    if dynamic_transform_found:
+        type_to_keep = 0
+    outer_transform.prune_dynamic_transform(type_to_keep)
+
+
+def _setup_transform_program(
+    user_transform_program: TransformProgram,
+    device: "qml.devices.Device",
+    resolved_execution_config: "qml.devices.ExecutionConfig",
+    cache=None,
+    cachesize=10000,
+) -> tuple[TransformProgram, TransformProgram, "qml.devices.ExecutionConfig"]:
+
+    device_transform_program, config = device.preprocess(execution_config=resolved_execution_config)
+
+    full_transform_program = qml.transforms.core.TransformProgram(user_transform_program)
+    inner_transform_program = qml.transforms.core.TransformProgram()
+
+    # Add the gradient expand to the program if necessary
+    if getattr(config.gradient_method, "expand_transform", False):
+        full_transform_program.add_transform(
+            qml.transform(config.gradient_method.expand_transform),
+            **config.gradient_keyword_arguments,
+        )
+    if config.use_device_gradient:
+        full_transform_program += device_transform_program
+    else:
+        inner_transform_program += device_transform_program
+
+    # Making sure dynamic_one_shot occurs at most once between the inner and outer transform programs
+    _prune_dynamic_transform(full_transform_program, inner_transform_program)
+
+    # If caching is desired but an explicit cache is not provided, use an ``LRUCache``.
+    if cache is True:
+        cache = LRUCache(maxsize=cachesize)
+        setattr(cache, "_persistent_cache", False)
+
+    # Ensure that ``cache`` is not a Boolean to simplify downstream code.
+    elif cache is False:
+        cache = None
+
+    # changing this set of conditions causes a bunch of tests to break.
+    no_interface_boundary_required = config.interface is None or config.gradient_method in {
+        None,
+        "backprop",
+    }
+    device_supports_interface_data = no_interface_boundary_required and (
+        config.interface is None
+        or config.gradient_method == "backprop"
+        or getattr(device, "short_name", "") == "default.mixed"
+    )
+    numpy_only = not device_supports_interface_data
+    if numpy_only:
+        inner_transform_program.add_transform(qml.transforms.convert_to_numpy_parameters)
+    if cache is not None:
+        inner_transform_program.add_transform(qml.workflow._cache_transform, cache=cache)
```

Comment:
Can we instead do `from ._cache_transform import _cache_transform` at the top of the file?

Reply:
Addressed in latest commit 👍🏼 .